File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/date/SetHoursTest.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class SetHoursTest {

    DateFunctionRepository dfr = new DateFunctionRepository();

    ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
            new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

    String message = "Invalid ISO 8601 Date format.";

    @Test
    void test() {

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("abcd"), "hourValue", new JsonPrimitive(121312)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage(message)
                .verify();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("202312=12"), "hourValue", new JsonPrimitive(1245)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectError()
                .verify();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("2053-10-04T14:10:50.700+00:00"), "hourValue",
                new JsonPrimitive(-12)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage("Hours should be in the range of 0 and 23")
                .verify();

        fep.setArguments(
                Map.of("isoDate", new JsonPrimitive("2023-10-19T23:44:11.615Z"), "hourValue",
                        new JsonPrimitive(24)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage("Hours should be in the range of 0 and 23")
                .verify();
    }

    @Test
    void validTest() {

        fep.setArguments(
                Map.of("isoDate", new JsonPrimitive("2023-10-19T06:44:11.615-11:11"), "hourValue",
                        new JsonPrimitive(-0)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("2023-10-19T00:44:11.615-11:11"))
                .verifyComplete();

        fep.setArguments(
                Map.of("isoDate", new JsonPrimitive("+071994-10-24T14:05:30.406+18:00"), "hourValue",
                        new JsonPrimitive(18)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString()
                                .equals("+071994-10-24T18:05:30.406+18:00"))
                .verifyComplete();

        fep.setArguments(
                Map.of("isoDate", new JsonPrimitive("-011994-10-24T14:05:30.406-18:00"), "hourValue",
                        new JsonPrimitive(17)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString()
                                .equals("-011994-10-24T17:05:30.406-18:00"))
                .verifyComplete();

        fep.setArguments(
                Map.of("isoDate", new JsonPrimitive("1994-10-24T14:05:30.406-18:00"), "hourValue",
                        new JsonPrimitive(10)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetHours")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("1994-10-24T10:05:30.406-18:00"))
                .verifyComplete();

    }
}

File: kirun-js/__tests__/engine/function/system/date/SetHoursTest.ts
import { KIRunFunctionRepository, KIRunSchemaRepository, Namespaces } from '../../../../../src';
import { DateFunctionRepository } from '../../../../../src/engine/function/system/date/DateFunctionRepository';
import { GetTimeZoneOffset } from '../../../../../src/engine/function/system/date/GetTimeZoneOffset';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

const dateFunctionRepo = new DateFunctionRepository();

test('testing SetHoursFunction', async () => {
    let setHours = await dateFunctionRepo.find(Namespaces.DATE, 'setHours');
    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    if (!setHours) {
        throw new Error('Function not available');
    }

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-10-04T11:45:38.939Z'],
            ['hoursValue', -109],
        ]),
    );

    expect(async () =>
        (await setHours?.execute(fep))?.allResults()[0].getResult().get('hour'),
    ).rejects.toThrowError('Hours should be in the range of 0 and 23');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-09-03T17:35:17.000Z'],
            ['hoursValue', 100],
        ]),
    );

    expect(async () =>
        (await setHours?.execute(fep))?.allResults()[0].getResult().get('hour'),
    ).rejects.toThrowError('Hours should be in the range of 0 and 23');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1970-01-20T15:58:57.561Z'],
            ['hoursValue', 1000],
        ]),
    );

    expect(async () =>
        (await setHours?.execute(fep))?.allResults()[0].getResult().get('hour'),
    ).rejects.toThrowError('Hours should be in the range of 0 and 23');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-10-19T06:44:11.615Z'],
            ['hoursValue', -0],
        ]),
    );

    expect((await setHours.execute(fep)).allResults()[0].getResult().get('hours')).toBe(0);

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-10-24T14:10:30.700+12:00'],
            ['hoursValue', 100],
        ]),
    );

    expect(async () =>
        (await setHours?.execute(fep))?.allResults()[0].getResult().get('hour'),
    ).rejects.toThrowError('Hours should be in the range of 0 and 23');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1994-10-24T14:05:30.406-18:00'],
            ['hoursValue', -100],
        ]),
    );

    expect(async () =>
        (await setHours?.execute(fep))?.allResults()[0].getResult().get('hour'),
    ).rejects.toThrowError('Hours should be in the range of 0 and 23');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1300-10-25T05:42:10.435+14:00'],
            ['hoursValue', -10000],
        ]),
    );

    expect(async () =>
        (await setHours?.execute(fep))?.allResults()[0].getResult().get('hour'),
    ).rejects.toThrowError('Hours should be in the range of 0 and 23');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1994-10-24T14:05:30.406-18:00'],
            ['hoursValue', 10],
        ]),
    );

    expect((await setHours.execute(fep)).allResults()[0].getResult().get('hours')).toBe(10);
});
